File: Tyche.Android/ui/src/main/java/com/felipearpa/ui/ViewState.kt
package com.felipearpa.ui

sealed class ViewState<out T : Any> {

    object Initial : ViewState<Nothing>()

    object Loading : ViewState<Nothing>()

    data class Success<T : Any>(private val value: T) : ViewState<T>() {
        operator fun invoke(): T = value
    }

    data class Failure(private val throwable: Throwable) : ViewState<Nothing>() {
        operator fun invoke(): Throwable = throwable
    }
}

fun <T : Any> ViewState<T>.isInitial() = this is ViewState.Initial

fun <T : Any> ViewState<T>.isLoading() = this is ViewState.Loading

fun <T : Any> ViewState<T>.isSuccess() = this is ViewState.Success

fun <T : Any> ViewState<T>.isFailure() = this is ViewState.Failure

inline fun <T : Any> ViewState<T>.onInitial(block: () -> Unit): ViewState<T> {
    if (isInitial()) {
        block()
    }
    return this
}

inline fun <T : Any> ViewState<T>.onLoading(block: () -> Unit): ViewState<T> {
    if (isLoading()) {
        block()
    }
    return this
}

inline fun <T : Any> ViewState<T>.onSuccess(block: (value: T) -> Unit): ViewState<T> {
    if (isSuccess()) {
        block((this as ViewState.Success).invoke())
    }
    return this
}

inline fun <T : Any> ViewState<T>.onFailure(block: (throwable: Throwable) -> Unit): ViewState<T> {
    if (isFailure()) {
        block((this as ViewState.Failure).invoke())
    }
    return this
}

File: Tyche.iOS/UI/Sources/UI/ViewState.swift
public enum ViewState<Value> {
    case initial
    case loading
    case success(Value)
    case failure(Error)
}

public extension ViewState {
    @inlinable
    func isLoading() -> Bool {
        if case .loading = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isSuccess() -> Bool {
        if case .success = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isFailure() -> Bool {
        if case .failure = self {
            return true
        }
        return false
    }
    
    @inlinable
    func errorOrNull() -> Error? {
        if case .failure(let error) = self {
            return error
        }
        return nil
    }
}
